pkg/recorder/recorder.go

```diff
@@ -0,0 +1,475 @@
+		m.srcPort = 0xffff
+	}
+	if t.DstPort != 0 {
+		m.dstPort = 0xffff
+	}
+	if t.Proto != 0 {
+		m.proto = 0xff
+	}
+	copy(m.srcMask, t.SrcPrefix.Mask)
+	copy(m.dstMask, t.DstPrefix.Mask)
+	return m
+}
+
+func countMaskOnes(m RecorderMask) int {
+	ones := 0
+	onesSrc, _ := m.srcMask.Size()
+	onesDst, _ := m.dstMask.Size()
+	ones += onesSrc + onesDst
+	if m.srcPort == 0xffff {
+		ones += 16
+	}
+	if m.dstPort == 0xffff {
+		ones += 16
+	}
+	if m.proto == 0xff {
+		ones += 8
+	}
+	return ones
+}
+
+func hashMask(x *RecorderMask) string {
+	return fmt.Sprintf("%s/%s/%x/%x/%x",
+		x.srcMask.String(), x.dstMask.String(),
+		int(x.srcPort), int(x.dstPort), int(x.proto))
+}
+
+func hashTuple(x *RecorderTuple) string {
+	return fmt.Sprintf("%s/%s/%x/%x/%x",
+		x.SrcPrefix.String(), x.DstPrefix.String(),
+		int(x.SrcPort), int(x.DstPort), int(x.Proto))
+}
+
+func (r *Recorder) triggerDatapathRegenerate() error {
+	return nil
+}
+
+func (r *Recorder) triggerMapUpsert(ri *RecInfo, t *RecorderTuple) error {
+	return nil
+}
+
+func (r *Recorder) triggerMapDelete(ri *RecInfo, t *RecorderTuple) error {
+	return nil
+}
+
+func (r *Recorder) applyDatapath(regen bool) error {
+	for _, e := range r.queue.add {
+		r.triggerMapUpsert(r.queue.ri, e)
+	}
+	r.queue.add = []*RecorderTuple{}
+	for _, e := range r.queue.del {
+		r.triggerMapDelete(r.queue.ri, e)
+	}
+	r.queue.del = []*RecorderTuple{}
+	r.queue.ri = nil
+	if regen {
+		log.Debugf("Recorder Masks: %v", r.recMask)
+		return r.triggerDatapathRegenerate()
+	}
+	return nil
+}
+
+func queuePurge(q []*RecorderTuple, i int) []*RecorderTuple {
+	q[i] = q[len(q)-1]
+	return q[:len(q)-1]
+}
+
+func (r *Recorder) queueAddDatapathFilter(ri *RecInfo, i int) {
+	if r.queue.ri == nil {
+		r.queue.ri = ri
+	}
+	r.queue.add = append(r.queue.add, &ri.Filters[i])
+}
+
+func (r *Recorder) queueDelDatapathFilter(ri *RecInfo, i int) {
+	if r.queue.ri == nil {
+		r.queue.ri = ri
+	}
+	filter := &ri.Filters[i]
+	hash := hashTuple(filter)
+	for i, e := range r.queue.add {
+		if hashTuple(e) == hash {
+			r.queue.add = queuePurge(r.queue.add, i)
+			return
+		}
+	}
+	r.queue.del = append(r.queue.del, filter)
+}
+
+func (r *Recorder) deleteRecInfoLocked(ri *RecInfo, withID bool) bool {
+	triggerRegen := false
+	for i, filter := range ri.Filters {
+		mask := convertTupleToMask(filter)
+		maskHash := hashMask(&mask)
+		if rm, found := r.recMask[maskHash]; found {
+			rm.users--
+			if rm.users == 0 {
+				delete(r.recMask, maskHash)
+				triggerRegen = true
+			}
+		}
+		r.queueDelDatapathFilter(ri, i)
+	}
+	if withID {
+		delete(r.recByID, ri.ID)
+	}
+	return triggerRegen
+}
+
+func (r *Recorder) DeleteRecorder(id ID) (bool, error) {
+	r.Lock()
+	defer r.Unlock()
+	if recInfo, found := r.recByID[id]; found {
+		return true, r.applyDatapath(r.deleteRecInfoLocked(recInfo, true))
+	}
+	return false, nil
+}
+
+func (r *Recorder) createRecInfoLocked(ri *RecInfo, withID bool) bool {
+	if withID {
+		r.recByID[ri.ID] = ri
+	}
```

Comment:
nit: `withID` is never false here in the current version. Do we still need it?

Reply:
I had it for symmetry for both upsert/delete ops.